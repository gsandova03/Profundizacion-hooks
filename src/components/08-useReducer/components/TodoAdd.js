import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from '../../../hooks/useForm';

const TodoAdd = ({ addTodo }) => {

  const [ { description }, handleInputChange, reset ] = useForm({
    description: ''
  })

  const handleAddSubmit = (e) => {
    e.preventDefault();
    console.log( description )
    if( description.trim().length <= 1 ){
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      todo: description,
      done: false
    };

    addTodo( newTodo );
    reset();
  }

  return (
    <>
      <h4 className="text-center">Aregar Todo</h4>
      <hr/>
      <form onSubmit={ handleAddSubmit }>
        <input 
          type="text" 
          name="description" 
          className="form-control"
          placeholder="tarea...."
          value={ description }
          onChange={ handleInputChange }
        />
        <button 
          type="submit" 
          className="btn btn-primary"
        >
          Agregar
        </button>
      </form>
    </>
  )
}

TodoAdd.propTypes = {
  addTodo: PropTypes.func.isRequired
}

export default TodoAdd
