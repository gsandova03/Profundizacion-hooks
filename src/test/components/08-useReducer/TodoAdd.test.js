import React from 'react'
import { shallow } from 'enzyme'
import TodoAdd from '../../../components/08-useReducer/components/TodoAdd'


describe('Pruebas en el componente <TodoAdd />', () => {
  
  const addTodo = jest.fn()
  const wrapper = shallow( <TodoAdd addTodo={ addTodo }/> )

  test('debe de mostrarse correctamente ', () => {


    expect( wrapper ).toMatchSnapshot();
  })

  test('No debe de llamar addTodo', () => {
    
    const formSubmit = wrapper.find('form').prop('onSubmit')
    
    formSubmit({ preventDefault(){} })
    expect( addTodo ).toHaveBeenCalledTimes(0)

  })
  
  test('debe de llamar la funcion addTodo', () => {
    
    const value = 'aprender React'
    //Simulamos el cambio de la informacion del input
    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description'
      }
    })

    //crear la referencia a nuesto formulario y la prop onSubmit
    const formSubmit = wrapper.find('form').prop('onSubmit')
    
    //Ejecutamos el submit para el 
    formSubmit({ preventDefault(){} })
    //evaluar el llamado de la funcion
    expect( addTodo ).toHaveBeenCalledTimes(1)
    expect( addTodo ).toHaveBeenCalledWith( expect.any(Object) )
    expect( addTodo ).toHaveBeenCalledWith({
      //Para este caso el id no es tenido en cuenta, lo unico importante es que sea de tipo number
      id: expect.any(Number),
      todo: value,
      done: false
    })

    //Al ejecutar el submit del formulario evualuamos que la funcion reset establesca nuestro valor por defecto, en este caso un '' string vacio
    expect( wrapper.find('input').prop('value') ).toBe('')

  })
  
})
