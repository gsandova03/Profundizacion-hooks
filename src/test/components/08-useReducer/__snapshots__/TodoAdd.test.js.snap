// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoAdd /> debe de mostrarse correctamente  1`] = `
<Fragment>
  <h4
    className="text-center"
  >
    Aregar Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="tarea...."
      type="text"
      value=""
    />
    <button
      className="btn btn-primary"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
